File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/aop/support/AnnotationMethodInterceptor.java
package org.xyattic.eventual.consistency.support.core.aop.support;

import org.aopalliance.intercept.MethodInterceptor;
import org.aopalliance.intercept.MethodInvocation;
import org.aspectj.lang.ProceedingJoinPoint;
import org.springframework.aop.ProxyMethodInvocation;
import org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint;
import org.springframework.core.Ordered;
import org.springframework.core.ResolvableType;
import org.springframework.core.annotation.AnnotatedElementUtils;
import org.xyattic.eventual.consistency.support.core.provider.PendingMessageContextHolder;

import java.lang.annotation.Annotation;
import java.util.Objects;

/**
 * @author wangxing
 * @create 2020/11/13
 */
public abstract class AnnotationMethodInterceptor<T extends Annotation> implements MethodInterceptor, Ordered {

    protected final Class<T> annotationType;

    public Class<T> getAnnotationType() {
        return annotationType;
    }

    public AnnotationMethodInterceptor() {
        ResolvableType[] resolvableTypes = ResolvableType.forClass(getClass())
                .as(AnnotationMethodInterceptor.class)
                .getGenerics();
        this.annotationType = (Class<T>) resolvableTypes[0].getRawClass();
    }

    @Override
    public Object invoke(MethodInvocation invocation) throws Throwable {
        try {
            if (!(invocation instanceof ProxyMethodInvocation)) {
                throw new IllegalStateException("MethodInvocation is not a Spring " +
                        "ProxyMethodInvocation: " + invocation);
            }
            ProxyMethodInvocation pmi = (ProxyMethodInvocation) invocation;
            ProceedingJoinPoint pjp = lazyGetProceedingJoinPoint(pmi);
            T annotation = getAnnotation(invocation);

            return doInvoke(annotation, pjp);
        } finally {
            PendingMessageContextHolder.clear();
        }
    }

    protected abstract Object doInvoke(T annotation, ProceedingJoinPoint pjp);

    /**
     * Return the ProceedingJoinPoint for the current invocation,
     * instantiating it lazily if it hasn't been bound to the thread already.
     *
     * @param rmi the current Spring AOP ReflectiveMethodInvocation,
     *            which we'll use for attribute binding
     * @return the ProceedingJoinPoint to make available to advice methods
     */
    protected ProceedingJoinPoint lazyGetProceedingJoinPoint(ProxyMethodInvocation rmi) {
        return new MethodInvocationProceedingJoinPoint(rmi);
    }

    protected T getAnnotation(MethodInvocation invocation) {
        T annotation =
                AnnotatedElementUtils.getMergedAnnotation(invocation.getMethod(), annotationType);
        if (annotation == null) {
            annotation =
                    AnnotatedElementUtils.getMergedAnnotation(invocation.getMethod().getDeclaringClass(), annotationType);
        }
        return Objects.requireNonNull(annotation, "cannot found @" + annotationType.getName() +
                " info");
    }

}

File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/aop/support/AnnotationMethodInterceptor.kt
package org.xyattic.eventual.consistency.support.core.aop.support

import org.aopalliance.intercept.MethodInterceptor
import org.aopalliance.intercept.MethodInvocation
import org.aspectj.lang.ProceedingJoinPoint
import org.springframework.aop.ProxyMethodInvocation
import org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint
import org.springframework.core.Ordered
import org.springframework.core.ResolvableType
import org.springframework.core.annotation.AnnotatedElementUtils

/**
 * @author wangxing
 * @create 2020/11/13
 */
abstract class AnnotationMethodInterceptor<T : Annotation> : MethodInterceptor, Ordered {

    val annotationType: Class<T>

    override fun invoke(invocation: MethodInvocation): Any? {
        check(invocation is ProxyMethodInvocation) {
            "MethodInvocation is not a Spring " +
                    "ProxyMethodInvocation: " + invocation
        }
        val pjp = lazyGetProceedingJoinPoint(invocation)
        val annotation = getAnnotation(invocation)
        return doInvoke(annotation, pjp)
    }

    protected abstract fun doInvoke(annotation: T, pjp: ProceedingJoinPoint): Any?

    /**
     * Return the ProceedingJoinPoint for the current invocation,
     * instantiating it lazily if it hasn't been bound to the thread already.
     *
     * @param rmi the current Spring AOP ReflectiveMethodInvocation,
     * which we'll use for attribute binding
     * @return the ProceedingJoinPoint to make available to advice methods
     */
    protected fun lazyGetProceedingJoinPoint(rmi: ProxyMethodInvocation): ProceedingJoinPoint {
        return MethodInvocationProceedingJoinPoint(rmi)
    }

    protected fun getAnnotation(invocation: MethodInvocation): T {
        var annotation = AnnotatedElementUtils.getMergedAnnotation(invocation.method, annotationType)
        annotation = annotation
                ?: AnnotatedElementUtils.getMergedAnnotation(invocation.method.declaringClass, annotationType)
        check(annotation != null) {
            "cannot found @${annotationType.name} info"
        }
        return annotation
    }

    init {
        val resolvableTypes = ResolvableType.forClass(javaClass)
                .`as`(AnnotationMethodInterceptor::class.java)
                .generics
        annotationType = resolvableTypes[0].rawClass as Class<T>
    }

}